src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CollectionInitializer.cs

```diff
@@ -3,68 +3,104 @@
             _cosmosDataStoreConfiguration = cosmosDataStoreConfiguration;
-            _initialCollectionThroughput = initialCollectionThroughput;
             _upgradeManager = upgradeManager;
+            _retryExceptionPolicyFactory = retryExceptionPolicyFactory;
+            _clientTestProvider = clientTestProvider;
             _logger = logger;
         }
 
         public async Task<Container> InitializeCollection(CosmosClient client)
         {
             Database database = client.GetDatabase(_cosmosDataStoreConfiguration.DatabaseId);
-            Container containerClient = database.GetContainer(_collectionId);
+            Container containerClient = database.GetContainer(_cosmosCollectionConfiguration.CollectionId);
 
-            _logger.LogInformation("Finding Container: {collectionId}", _collectionId);
-            var existingContainer = await database.TryGetContainerAsync(_collectionId);
+            _logger.LogInformation("Finding Container: {collectionId}", _cosmosCollectionConfiguration.CollectionId);
 
-            if (existingContainer == null)
+            AsyncPolicy retryPolicy = _retryExceptionPolicyFactory.GetRetryPolicy();
+
+            ContainerResponse container = await retryPolicy.ExecuteAsync(async () =>
             {
-                _logger.LogInformation("Creating Cosmos Container if not exits: {collectionId}", _collectionId);
+                var existingContainer = await database.TryGetContainerAsync(_cosmosCollectionConfiguration.CollectionId);
+                if (existingContainer == null)
+                {
+                    _logger.LogInformation("Creating Cosmos Container if not exits: {collectionId}", _cosmosCollectionConfiguration.CollectionId);
+
+                    var containerResponse = await database.CreateContainerIfNotExistsAsync(
+                        _cosmosCollectionConfiguration.CollectionId,
+                        $"/{KnownDocumentProperties.PartitionKey}",
+                        _cosmosCollectionConfiguration.InitialCollectionThroughput);
+
+                    containerResponse.Resource.DefaultTimeToLive = -1;
 
-                var containerResponse = await database.CreateContainerIfNotExistsAsync(
-                    _collectionId,
-                    $"/{KnownDocumentProperties.PartitionKey}",
-                    _initialCollectionThroughput);
+                    existingContainer = await containerClient.ReplaceContainerAsync(containerResponse);
 
-                containerResponse.Resource.DefaultTimeToLive = -1;
+                    if (_cosmosCollectionConfiguration.InitialCollectionThroughput.HasValue)
+                    {
+                        ThroughputProperties throughputProperties = ThroughputProperties.CreateManualThroughput(_cosmosCollectionConfiguration.InitialCollectionThroughput.Value);
+                        await containerClient.ReplaceThroughputAsync(throughputProperties);
```

Comment:
Do you think this entire block should be retried or each individual call?
Assuming something fails it would start again be reading the collection, if it previously failed on `ReplaceThroughputAsync`, this setting might not be applied. Thoughts?

Reply:
I guess putting retries around each statement lowers the chances that the throughput is not applied...